src/pin.cpp

```diff
@@ -0,0 +1,72 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cstdint>
+
+#include <libhal-stm32f4/pin.hpp>
+
+#include <libhal-util/bit.hpp>
+#include <libhal/units.hpp>
+
+#include "gpio_reg.hpp"
+
+namespace hal::stm32f4 {
+
+const pin& pin::function(hal::stm32f4::pin_function p_function) const
+{
+  auto port_reg = get_reg(m_port);
+  auto pin_mode_mask =
+    hal::bit_mask::from<2 * m_pin, 2 * m_pin + 1>(port_reg->pin_mode);
+  uint8_t alt_func = 0;
```

Comment:
```suggestion
    hal::bit_mask::from<2 * m_pin, 2 * m_pin + 1>(port_reg->pin_mode);
  uint8_t alt_func = 0;
```

This shouldn't work because m_pin is a runtime value and templates need compile time constants.